### PR TITLE
add OpenSSL provider for ESDM

### PR DESCRIPTION
**Pull request by thillux**
This adds two OpenSSL providers:
- a seed source
- a full RNG

Writing an OpenSSL 1.x engine did not made sense for me, as OpenSSL 1.1 is only supported until Sep. 23.

### CODE REVIEW

File: frontends/openssl-provider/common.c [hunk 1]
```diff
@@ -0,0 +1,406 @@
+#include "common.h"
```
Comment (smuellerDD) [inline frontends/openssl-provider/common.c:1]: Please add the comment header

File: frontends/openssl-provider/common.c [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+#include "esdm_rpc_client.h"
+#include "helper.h"
+#include "math_helper.h"
+#include "openssl/core_names.h"
+#include "openssl/crypto.h"
+#include "openssl/evp.h"
+#include "openssl/params.h"
+#include "visibility.h"
+
+#include <errno.h>
```
Comment (smuellerDD) [inline frontends/openssl-provider/common.c:12]: Nit: may I ask to move the external headers to the top?

File: frontends/openssl-provider/common.c [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+}
+
+static void esdm_rand_freectx(void *ctx)
+{
+  struct esdm_rand_ctx *rand = ctx;
+
+  if (rand == NULL)
+    return;
+
+  CRYPTO_THREAD_lock_free(rand->lock);
```
Comment (smuellerDD) [inline frontends/openssl-provider/common.c:66]: Sanity check: rand->lock may be NULL, can this OpenSSL handle NULL?

File: frontends/openssl-provider/common.c [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+static size_t esdm_rand_nonce(void *ctx __unused, unsigned char *out,
+                              unsigned int outlen, size_t min_noncelen __unused,
+                              size_t max_noncelen __unused)
+{
+  ssize_t ret;
+
+  if (out == NULL)
+    return outlen;
+
+  esdm_invoke(esdm_rpcc_get_random_bytes_full(out, outlen));
```
Comment (smuellerDD) [inline frontends/openssl-provider/common.c:140]: Perhaps get_random_bytes_min would suffice?

File: frontends/openssl-provider/common.c [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+    uint64_t len;
+    uint64_t entropyBits;
+    uint8_t buf[2048] /* should be large enough for entropy from all active
+                         sources */
+  } __attribute__((__packed__));
+  size_t buffer_size = sizeof(struct esdm_seed_buffer);
+  struct esdm_seed_buffer *seed_buf;
+  ssize_t ret;
+
+  *buffer = OPENSSL_secure_zalloc(sizeof(struct esdm_seed_buffer));
```
Comment (smuellerDD) [inline frontends/openssl-provider/common.c:164]: Wouldn't it make sense to move the alloc call below the sanity checks?

Comment (smuellerDD) [inline frontends/openssl-provider/common.c:164]: Shouldn't the returned data only hold the seed_buf and not the full structure? The caller may have no notion about the seed_buf structure. But I admit that it does not hurt to also return the additional information.

File: frontends/openssl-provider/common.c [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+    if (ret != 0)
+      goto err;
+  }
+
+  esdm_invoke(esdm_rpcc_get_seed(*buffer, buffer_size, 0));
+  if (ret <= 0)
+    goto err;
+
+  seed_buf = (struct esdm_seed_buffer *)*buffer;
+  if (seed_buf->entropyBits < (uint64_t)entropy)
```
Comment (smuellerDD) [inline frontends/openssl-provider/common.c:182]: Sanity check: are you sure that the OpenSSL entropy is measured in bits here? If yes, may I ask you to rename the variable to entropy_bits.

Side note: I lost days in debugging issues of bits vs bytes. Thus, I made a habit now to mark all variables that hold bit sizes.

File: frontends/openssl-provider/common.h [hunk 7]
```diff
@@ -0,0 +1,27 @@
+#ifndef ESDM_OPENSSL_RNG_PROVIDER_COMMON_H
```
Comment (smuellerDD) [inline frontends/openssl-provider/common.h:1]: Please add the header comments

File: frontends/openssl-provider/rng-provider.c [hunk 8]
```diff
@@ -0,0 +1,10 @@
+#include "common.h"
```
Comment (smuellerDD) [inline frontends/openssl-provider/rng-provider.c:1]: Please add header comments

File: frontends/openssl-provider/seed-src-provider.c [hunk 9]
```diff
@@ -0,0 +1,7 @@
+#include "common.h"
```
Comment (smuellerDD) [inline frontends/openssl-provider/seed-src-provider.c:1]: Please add header comments.

File: meson_options.txt [hunk 10]
```diff
@@ -340,6 +340,9 @@ Using CUSE, the device files are provided.
 option('linux-getrandom', type: 'feature', value: 'enabled',
        description: 'Enable the Linux getrandom system call support.')
 
+option('openssl-rand-provider', type: 'feature', value: 'disabled',
+       description: 'Enable the OpenSSL >= 3 RAND/SEED-SRC provider support.')
```
Comment (smuellerDD) [inline meson_options.txt:344]: Can you please add additional hints that this is a frontend, no backend?

**Review by smuellerDD — CHANGES_REQUESTED** [review]

File: tests/openssl-rand-provider/env.c [hunk 11]
```diff
@@ -0,0 +1,99 @@
+/*
```
Comment (smuellerDD) [inline tests/openssl-rand-provider/env.c:1]: Note to self: this code duplication of env.c is not good, I need to think about how to handle it. This comment, however, shall not hold up any merging.

File: tests/openssl-rand-provider/openssl_rng_tester.c [hunk 12]
```diff
@@ -0,0 +1,103 @@
+#include <openssl/provider.h>
```
Comment (smuellerDD) [inline tests/openssl-rand-provider/openssl_rng_tester.c:1]: Please add comment header

File: tests/openssl-rand-provider/openssl_rng_tester.c [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+#include <assert.h>
+#include <stdbool.h>
+#include <string.h>
+
+#include "env.h"
+
+static void test_random() {
+    unsigned char bytes[300];
+    int ret = RAND_bytes(bytes, 300);
+    fprintf(stderr, "%i\n", ret);
```
Comment (smuellerDD) [inline tests/openssl-rand-provider/openssl_rng_tester.c:16]: nit: coding style: I very much like KNF which asks for a blank line between variable declarations and code.

File: tests/openssl-rand-provider/openssl_rng_tester.c [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        test_random();
+    if (strncmp(test, "instantiate_pr", strlen("instantiate_pr")) == 0)
+        test_instantiate(true);
+    if (strncmp(test, "instantiate_full", strlen("instantiate_full")) == 0)
+        test_instantiate(false);
+}
+
+int main(int argc, char **argv)
+{
+    assert(argc == 4);
```
Comment (smuellerDD) [inline tests/openssl-rand-provider/openssl_rng_tester.c:86]: coding style: KNF asks for variable declarations first followed by code. Please move the assert down.

File: tests/openssl-rand-provider/openssl_rng_tester.c [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+    char *provider_search_path = argv[1];
+    char *test = argv[2];
+    char *type = argv[3];
+
+    int ret = env_init();
+    if (ret)
+        return ret;
+
+    ret = OSSL_PROVIDER_set_default_search_path(NULL, provider_search_path);
+    assert(ret == 1);
```
Comment (smuellerDD) [inline tests/openssl-rand-provider/openssl_rng_tester.c:96]: assert here implies that env_fini is not called. Unless you have a very important reason, may I ask to "manually" check ret, and return ret?

**Review by smuellerDD — CHANGES_REQUESTED** [review]

**Comment by thillux** [comment]
I'll started working on your suggestions, but I am not able to finish this today. I'll notify you, when done.

File: frontends/openssl-provider/common.c (hunk 3 above)
Comment (thillux) [reply frontends/openssl-provider/common.c:83]: At least in the version of OpenSSL I looked at (3.1.1):

```c
void CRYPTO_THREAD_lock_free(CRYPTO_RWLOCK *lock)
{
    if (lock == NULL)
        return;

# ifdef USE_RWLOCK
    pthread_rwlock_destroy(lock);
# else
    pthread_mutex_destroy(lock);
# endif
    OPENSSL_free(lock);

    return;
}
```

File: frontends/openssl-provider/common.c (hunk 4 above)
Comment (thillux) [reply frontends/openssl-provider/common.c:157]: I think get_random_bytes_min should also be sufficient here, as this is only used as a nonce additional to entropy, when instantiating DRBGs in OpenSSL.

File: frontends/openssl-provider/common.c (hunk 6 above)
Comment (thillux) [reply frontends/openssl-provider/common.c:199]: I'm pretty sure, that the entropy argument uses bits. I'll rename the argument accordingly. It is stated in the documentation and I also found it in the OpenSSL source code, e.g.:

```c
static size_t seed_get_seed(void *vseed, unsigned char **pout,
                            int entropy, size_t min_len, size_t max_len,
                            int prediction_resistance,
                            const unsigned char *adin, size_t adin_len)
{
    size_t bytes_needed;
    unsigned char *p;

    /*
     * Figure out how many bytes we need.
     * This assumes that the seed sources provide eight bits of entropy
     * per byte.  For lower quality sources, the formula will need to be
     * different.
     */
    bytes_needed = entropy >= 0 ? (entropy + 7) / 8 : 0;
    if (bytes_needed < min_len)
        bytes_needed = min_len;
    if (bytes_needed > max_len) {
        ERR_raise(ERR_LIB_PROV, PROV_R_ENTROPY_SOURCE_STRENGTH_TOO_WEAK);
        return 0;
    }

    p = OPENSSL_secure_malloc(bytes_needed);
    if (p == NULL)
        return 0;
    if (seed_src_generate(vseed, p, bytes_needed, 0, prediction_resistance,
                          adin, adin_len) != 0) {
        *pout = p;
        return bytes_needed;
    }
    OPENSSL_secure_clear_free(p, bytes_needed);
    return 0;
}
```

File: frontends/openssl-provider/common.c (hunk 5 above)
Comment (thillux) [reply frontends/openssl-provider/common.c]: I changed that accordingly. I just thought of the header values as additional data, that does not harm if later cryptographically hashed (and no entropy is accounted for it).

**Comment by thillux** [comment]
I think I'm done with the first round.

**Comment by smuellerDD** [comment]
All applied after some wrestling as it clashed with the Botan patches. Thanks.

However, I never asked to hijack your code. Thus, I am happy to receive a patch that changes the copyright owner to yourself in the file headers. :-)

**Pull request closed without merging**